### DISCUSSION TITLE
Remove the entire content of a folder in storage

### DISCUSSION THREAD

**Discussion by halindraprakoso** [supabase/supabase#4218] [Questions] [answered]
Hello,

is it possible to delete the folder or entire content of a folder using storage.remove()?

**Comment by GaryAustin1** [comment, marked as answer]
No.  You can empty a bucket with a call or use the UI.
 Otherwise you have to build up a list of file paths from the "folder" and remove them in a loop or as an array.  I'm not sure how well that would work for large amount  of files in a single call with an array, or looping in a database function.  You probably need to run such code in a cloud function or on your server if you have one.
The new studio open source likely has examples of deleting all files in a folder as the user interface does that for you.

**Reply by halindraprakoso** [reply]
yeah, I thought it would work with wildcard like this:
```
const { error: removeError } = await supabase.storage
			.from(bucket)
			.remove([`${user_id}/${folderName}/*`]);
```

I can do this instead:
```
const { data:list, error } = await supabase.storage.from(bucket).list(`${user_id}/${folderName}`);
const filesToRemove = list.map((x) => `${user_id}/${folderName}/${x.name}`);

const { data, error } = await supabase.storage.from(bucket).remove(filesToRemove);
```

**Reply by GaryAustin1** [reply]
I have never seen the * method work.   Be careful if you have a lot of files.  Besides server load, if .remove passes data in the url versus body like post, you will blow up url length quickly.  I don't know how Supabase storage API implements DELETE as body is optional.  Normally all resources are in the url for DELETE.

**Reply by halindraprakoso** [reply]
fortunately, I won't be dealing with a large number of files for this case but that's an interesting point to remember!

**Comment by logemann** [comment]
The UI can also delete complete folders... so do they also iterate through the files and then delete the folder itself?

**Reply by GaryAustin1** [reply]
Yes, last time I looked at the source code (6 months ago) that is what the UI does.  The folder is not a real thing in s3 storage.   In the UI it is just a placeholder object with a special name and used only to help display things.

**Reply by logemann** [reply]
yeah you are right. Just tested it and still need to iterate files. And of course, a folder is just a simple object in S3 and therefore same mechanics apply to supabase.

**Comment by bart** [comment]
Problem with the solution from above are folders inside folders. Supabase doesn't provide a way for recursive listing as far as I know so we have to implement it manually, right?

**Reply by GaryAustin1** [reply]
Not sure which solution.  If I were doing this myself I would probably use an rpc call and process the storage objects table directly and make http storage-api delete calls as needed.  OR just do it in server client code as you have access to the storage schema and can select files from storage.objects and loop thru deleting them as desired.

**Reply by bart** [reply]
Thanks for your reply! I mentioned the solution by @halindraprakoso He is listing all files in the given folder and maps the result to `filesToRemove` array. I'll check your way of accessing `storage` database directly. Nice approach imo.

**Reply by GaryAustin1** [reply]
And actually there is a bulk delete once you have the filepaths in the same bucket.  So you don't have to actually loop doing a delete per file..

**Reply by bart** [reply]
May I ask where the bulk delete is documented? If I do `await client.storage.from('<bucket-name>').remove(['<path-to-folder>'])` it doesn't get deleted

**Reply by GaryAustin1** [reply]
That is single or bulk.  You are probably not meeting RLS storage policy if it did not delete.

This has some examples of processing storage.objects in the database and doing deletes from there with both http and pg_net extensions.
https://github.com/GaryAustin1/supa-file-helper/blob/main/main.sql

**Reply by bart** [reply]
Unfortunately it looks like it doesn't work as bulk operation. I haven't set any RLS storage policy yet. Removing single files work but folders just won't get removed. That's weird, will look further into it.

**Reply by bart** [reply]
@GaryAustin1 Are you able to remove folders by simply calling `await client.storage.from('<bucket-name>').remove(['<path-to-folder>'])` ?

**Reply by GaryAustin1** [reply]
I'll check, There was just a big storage server change.  But you say you don't have RLS storage policy.  Then you can't delete at all so single should not work either.

**Reply by GaryAustin1** [reply]
Both single and bulk delete work for me.  With a delete policy in place.
There will be no error if you don't meet policy.  RLS is silent for everything but inserting.

```js
    const data = await supabases[2]
        .storage
        .from('testp')
        .remove(['1000.JPG','200.jpg'])
```

**Reply by bart** [reply]
But it does :) It's a public bucket but as far as I know it just means everyone is allowed to read. Anyways, using Supabase service role client does remove files in that bucket but not folders.

**Reply by bart** [reply]
I can see you are removing multiple files, that works. But can you remove a folder with sub-folders and files inside?

**Reply by GaryAustin1** [reply]
There is no such thing as a folder.  You have to remove files.

Folders are derived from the paths with string processing.  They don't exist as far as the API is concerned.  The UI does make up a fake folder for viewing.  But the API does not know about it or care.

Bulk delete is an array of ALL files you want deleted.  There is no delete files in this folder as there really is no folder.

**Reply by bart** [reply]
All right, got you. So I need to do the listing anyways. Will do it the DB way to fetch storage objects starting with a given path and then doing bulk delete on these objects. Thank you! I know that S3 also doesn't have a folders concept. They are just part of stored object locations. So ya, it's the same here because it's using S3 under the hood.

**Reply by GaryAustin1** [reply]
Correct.  And just processing the table directly seems pretty easy versus the list command.

**Reply by bart** [reply]
Indeed it is. That is a pretty good approach. Thanks again!

**Reply by Andriiklymiuk** [reply]
well, I came up with this solution for usage in edge function (not the best one, but does the job)
```
function chunkArray<T>(array: T[], size: number): T[][] {
  let result: T[][] = [];
  for (let i = 0; i < array.length; i += size) {
    result.push(array.slice(i, i + size));
  }
  return result;
}

async function listAllFiles(
  client: SupabaseClient,
  bucket: string,
  path: string,
): Promise<string[]> {
  let files: string[] = [];
  const { data, error } = await client.storage.from(bucket).list(path, {
    limit: 100,
  });
  if (error) throw new Error(`Failed to list files: ${error.message}`);
  console.log("Data: " + JSON.stringify(data, null, 2));

  for (const file of data) {
    const fullPath = path + file.name;
    if (!file.name.includes(".")) {
      const subFiles = await listAllFiles(client, bucket, fullPath + "/");
      files = files.concat(subFiles);
    } else {
      files.push(fullPath);
    }
  }
  return files;
}

async function deleteFilesInChunks(
  client: SupabaseClient,
  bucket: string,
  basePath: string,
) {
  try {
    const filesToDelete = await listAllFiles(client, bucket, basePath);
    if (filesToDelete.length > 0) {
      const chunkSize = 100;
      const fileChunks = chunkArray(filesToDelete, chunkSize);
      for (const chunk of fileChunks) {
        const { data: filesRemoved, error: filesRemovedError } = await client
          .storage
          .from(bucket)
          .remove(chunk);
        if (filesRemovedError) {
          console.error("Failed to remove files:", filesRemovedError.message);
          continue;
        }
        console.log("Files removed:", JSON.stringify(filesRemoved, null, 2));
      }
    } else {
      console.log("No files to delete.");
    }
  } catch (error) {
    console.error("An error occurred:", error.message);
  }
}

deleteFilesInChunks(
  supabaseClient,
  "my-target-bucket",
  "pathToStartDeletionInMyCaseIsUserId/",
);

```



I know it can be heavy for many files, but has someone came up with better alternative maybe for dealing with many files deletion? or I am missing smth and supabase already has some helper?

**Reply by AntonOfTheWoods** [reply]
> There is no such thing as a folder. You have to remove files.

Hmmm. I fully understand that, but if that is the case, why do they NOT give a listing of all files in the bucket, in spite of the docs ("List all files in a bucket"), when you try and get a listing? If I use `content_files = supabase.storage.from_("mybucket").list()`, and that "has a single directory with 1000 files" in it, I only get one single entry returned - the "directory". So in some places they treat them as you suggest (which represents the underlying storage) but elsewhere they want you to think they're "directories". And they still don't give you any way to recurse...

So we have to manually roll our own. At least it *appears* like we can easily tell a "directory" by having `metadata: null`, so I guess there's that?

**Comment by t6tv8e** [comment]
Would be very very nice if it's possible to remove items via a wildcard

**Reply by sanchezcarlosjr** [reply]
Even better if it supports glob or regex syntax.

**Comment by gyenabubakar** [comment]
I found a workaround that may help. Not an optimal solution but better than the most obvious approach.

Consider these two queries:
```ts
const { data: filePaths, error: filePathsError } = await this.supabase
  .schema("storage")
  .from("objects")
  .select("name")
  .eq("bucket_id", 'some-bucket-name')
  .like("name", `some-uuid-here%`); // remove or keep the % based on your needs

const { data: listedFiles } = await this.supabase.storage
  .from(bucket)
  .list(pathWithoutBucket);

console.log({ filePaths, listedFiles });
```

The result:
```json
{
    "filePaths": [
        {
            "name": "some-uuid-here/officers/officer-uuid-here/avatar/avatar.png"
        }
    ],
    "listedFiles": [
        {
            "name": "officers",
            "id": null,
            "updated_at": null,
            "created_at": null,
            "last_accessed_at": null,
            "metadata": null
        }
    ]
}
```

The first query returns the full path of all the objects in the folder `some-uuid-here`—it retrieves all subfolders and their folders and files and so on—whilst the second only returns the names of the files or folders directly in the `some-uuid-here` folder. So, if you want to delete files recursively, list them using the first query.

Then, you can delete the files using the paths from the first query:
```ts
const pathsToDelete = filePaths
      .filter((path): path is { name: string } => !!path.name)
      .map((f) => f.name);

await this.supabase.storage.from(bucket).remove(pathsToDelete);
```

---
If you're working locally, remember to expose the `storage` schema in your `supabase/config.toml` file. Also, define RLS policies on your buckets to ensure nothing is leaked wrongly.

---
---
---

Side note: This query deletes all files/folders recursively:
```ts
await this.supabase
      .schema('storage')
      .from('objects')
      .delete()
      .eq('bucket', 'some-bucket-name')
      .like('name', `some-folder-path%`);
```
...at least, that's what Supabase Studio shows: the files/folders no longer exist after running this query. I chose to go with the first approach because whilst the first approach truly deletes the files/folders, I dunno whether this query simply removes the database records or it actually deletes the files themselves as well. My guess is the former. Maybe a contributor can clarify?

***
***

Edit — stick with the first approach:

![Screenshot 2025-03-18 at 2 24 04 PM](https://github.com/user-attachments/assets/fb9f2b96-ec82-4885-9a36-66194ed12d5c)

**Comment by habibium** [comment]
I think this is the cleanest solution:
```ts
  const { data: userDocuments } = await supabase.storage.from(bucketName).listV2({ prefix: userId });  // or the path you want to remove
  const objectsToRemove = userDocuments?.objects.map((x) => x.name) || [];
  await supabase.storage.from(bucketName).remove(objectsToRemove);
  ```
